// bootstrap-colorselector/webpack.config.js
const path            = require("path");
const webpack         = require("webpack");
const UglifyJsPlugin  = webpack.optimize.UglifyJsPlugin;

let env         = process.env.WEBPACK_ENV;
let libraryName = "bootstrap-colorselector";
let outputFile;
let plugins = [];

if (env === 'build') {
  plugins.push(new UglifyJsPlugin({ minimize: true }));
  outputFile  = `${libraryName}.min.js`;
} else {
  outputFile  = `${libraryName}.js`;
}

const config = {
  entry: path.join(__dirname, "lib", "bootstrap-colorselector-0.2.0", "js/bootstrap-colorselector.js"),
  devtool: "source-map",
  output: {
    path: path.join(__dirname, "dist"),
    filename: outputFile,
    library: libraryName,
    libraryTarget: "umd",
    umdNamedDefine: true
  },
  module: {
    rules: [
      { test: /\.js$/, loader: "babel-loader", exclude: /node_modules/, options: { presets: ["es2015"] } },
      { test: /\.js$/, loader: "eslint-loader", exclude: /node_modules/ }
    ]
  },
  resolve: { modules: [path.join(__dirname, "src")], extensions: [".js"] },
  plugins: plugins
};

module.exports = config;
